fix: kubernetes and kyverno version annotations in kyverno-policies helm chart (#1165)

sergey198828 opened the pull request:
## Explanation

When deploy **kyverno-policies** helm chart it specifies versions as annotations of each policy:
 - **kyverno.io/kyverno-version**: Automatically detects installed version of kyverno. If version couldn't be detected value taken from Chart.yaml AppVersion field.
 - **kyverno.io/kubernetes-version**: Takes from values.yaml for which new key been added there.

## Related issue

/closes https://github.com/kyverno/policies/issues/1165
@realshuting

## Milestone of this PR
/milestone release-1.13

## Documentation (required for features)
My PR does not contain any new functionalities to Kyverno. Documentation for helm chart been updated in README.md and comments been added to values.yaml

## What type of PR is this
/kind bug

## Proposed Changes
This change maintain appropriate annotations to Kubernetes resources when deploy policy using kyverno-policies helm chart. It gives users control to which versions of Kubernetes policies suppose to run and which not.

### Proof Manifests
Not applicable since no change to Kyverno itself been made, only to helm chart.

Before change
```bash
helm template kyverno-policies ./charts/kyverno-policies | grep kubernetes-version
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"
    kyverno.io/kubernetes-version: "1.22-1.23"

helm template kyverno-policies ./charts/kyverno-policies | grep kyverno-version   
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
    kyverno.io/kyverno-version: 1.6.0
```
After change
```bash
helm template kyverno-policies ./charts/kyverno-policies | grep kubernetes-version
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30
    kyverno.io/kubernetes-version: 1.28-1.30

helm template kyverno-policies ./charts/kyverno-policies | grep kyverno-version
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
    kyverno.io/kyverno-version: v1.13.0-rc.1
```

## Checklist
- [ x ] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [ x ] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
- [ ] CLI support should be added and my PR doesn't contain that functionality.

realshuting commented:
Hi @sergey198828 - can you send the PR to `main` branch instead of release-1.13? We merge all PRs to main first then cherry-pick to release branches if needed.

sergey198828 commented:
> Hi @sergey198828 - can you send the PR to `main` branch instead of release-1.13? We merge all PRs to main first then cherry-pick to release branches if needed.

@realshuting PR been sent to main instead of  release-1.13

realshuting commented on charts/kyverno-policies/values.yaml:
```diff
@@ -114,3 +114,7 @@ skipBackgroundRequests: ~
+
+# -- Kubernetes version
+# Against which the policy should work. Ideally range of versions no more than two prior (ex., 1.28-1.31), must be enclosed in quotes
+kubernetesVersion: "1.28-1.31"
```
Why do we want a Helm option to configure it? It should be set as part of the release process, we can add a `testedKubernetesVersions: "1.28-1.31"` attribute in `charts/kyverno-policies/Chart.yaml`.

sergey198828 replied:
Helm options are convenient way to allow users customize behavior of deployments.

Since this value is propagated to manifests and effectively impact behavior of deployment I thought it was a good idea to let users customize this field if they are running older or newer versions of Kubernetes without need to modify Chart files.

Otherwise, like it happened in our organization for example, we deployed policies v3.2.5 to newer version of Kubernetes 1.28 and found out they are not effective, since manifests has hardcoded versions 1.22-1.23. And there is no way to modify this without pulling and changing chart.

sergey198828 replied:
Okay, maybe I misunderstood.

I just have checked latest version of Kyverno Policies chart 3.3.0 and I still see it has annotation kyverno.io/kubernetes-version: "1.22-1.23". Effectively this means even though I applied policies to my clusters they are not working.

Current K8s versions are 1.29-1.31 and even 1.28 is end of life already.

I have to change hardcoded templates myself to make this annotation match versions of my clusters.

I thought it would be good idea to make it possible for users customize while still keep default value match the officially tested Kubernernetes versions.

Release team will still need record this value in default values.yaml instead of tens of templates.

realshuting replied:
I see those are two different configurations. One is mandatory for the release team to update the officially tested Kubernetes versions, and another is an optional configuration offered to users to customize this value.

sergey198828 replied:
Totally makes scenes. I have updated Pull Request. Release team supposed to ensure values in Chart.yaml are updated and templates use them by default. Optional parameter added to Values in case user want to customize it. Please review.

realshuting reviewed (changes requested):
Sorry for the delay @sergey198828 , I added a comment regarding a chart value, please see below.

realshuting commented:
Can you fix the code-gen failure by running `make codegen-helm-all`?

sergey198828 commented:
**Re-organized branch:**
1. Rebased from main
2. Squashed all my commits into one and signed off

**Fixes:**
As we agreed, by default Kyverno policies chart will take supported Kubernetes version from Chart,yaml kubeVersion field which is in responsibility of release team to maintain. Optionally user can override this behavior using values file.
Kyverno version will be by default taken from Chart,yaml appVersion field, installation will attempt to detect Kyverno which already installed to cluster and match version.

realshuting commented on charts/kyverno-policies/templates/restricted/restrict-volume-types.yaml:
```diff
@@ -16,8 +16,8 @@ metadata:
     policies.kyverno.io/minversion: 1.6.0
-    kyverno.io/kubernetes-version: "1.22-1.23"
-    kyverno.io/kyverno-version: 1.6.0
+    kyverno.io/kubernetes-version: "{{ default .Chart.KubeVersion .Values.kubeVersionOverride }}"
```
Need to automate this as part of the release process.

realshuting replied:
Track via https://github.com/kyverno/kyverno/issues/11630.

realshuting reviewed (approved):
/lgtm

Thanks for the contribution @sergey198828 !
